src/main.rs

```diff
@@ -782,6 +794,79 @@ fn start_vmm(cmd_arguments: ArgMatches) -> Result<Option<String>, Error> {
     r.map(|_| api_socket_path)
 }
 
+// This is a best-effort solution to the latency induced by the RCU
+// synchronization that happens in the kernel whenever the file
+// descriptor table is full and needs to be expanded. The initial table
+// has 64 entries on amd64.
+// Rust programs that use {File,EventFd}::try_clone() to share handles
+// are impacted by this issue.
+// This behavior is quite noticeable in the restore snapshot scenario,
+// the latency is a big chunk of the total time required to start
+// cloud-hypervisor and to restore the snapshot.
+// The kernel has an optimization in the expand table code path where it
+// doesn't call synchronize_rcu() if there is only one thread in the
+// process.
+// To address this issue we attempt to expand the descriptor table at
+// application start, when it has only one thread.
+fn expand_fdtable() -> Result<(), FdTableError> {
+    let mut limits = libc::rlimit {
+        rlim_cur: 0,
+        rlim_max: 0,
+    };
+
+    // SAFETY: FFI call with valid arguments
+    if unsafe { libc::getrlimit(libc::RLIMIT_NOFILE, &mut limits) } < 0 {
+        return Err(FdTableError::GetRLimit(io::Error::last_os_error()));
+    }
+
+    let mut table_size: i32;
+
+    if limits.rlim_cur < 4096 {
+        table_size = limits.rlim_cur as i32;
+
+        let new_limits = libc::rlimit {
+            rlim_cur: std::cmp::min(4096, limits.rlim_max),
+            rlim_max: limits.rlim_max,
+        };
+
+        // If this fails we are using the current rlim_cur which is lower
+        // than 4096.
+        // SAFETY: FFI call with valid arguments
+        if unsafe { libc::setrlimit(libc::RLIMIT_NOFILE, &new_limits) } == 0 {
+            table_size = new_limits.rlim_cur as i32;
+        }
+    } else {
+        table_size = 4096;
+    }
+
+    if table_size <= 3 {
+        return Ok(());
```

Comment:
Why return if the table is smaller than 3?

Reply:
The first three file handles are stdin, stdout and stderr. We don't want to mess with those. Also if the table_size is 3 it means we didn't manage to set the soft limit (rlim_cur) to 4096. Either our call failed or the hard limit is <= 3. Either way there is nothing we can possibly do, most likely the application won't even run.

Reply:
Right, makes sense.  Could that be explained in a comment?

Reply:
Comment added.